utils/uri/matcher.go

```diff
@@ -0,0 +1,75 @@
+package uri
+
+import (
+	"errors"
+	"strings"
+)
+
+var (
+	ErrInvalidURI = errors.New("invalid URI string")
+)
+
+//hs:<service_name>:<subsystem>:<version>:<category>:<resource_name>?<allowed_arguments>#<permission_metadata>
+type URI struct {
+	Path      string
+	Arguments map[string]string
+	Metadata  map[string]string
```

Comment:
The URIs should ideally be immutable so I'd say it'd be a good idea to keep these fields private.